### PR TITLE
Cesmdev update

### PR DESCRIPTION
**Pull request by mvertens**
Updates from cesmdev master

1) Add in option for CLMGSWP3 for datm. Also add in capability for driver
namelist generation to merge CAM and CLM drv_flds_in namelists
together, and abort on a conflict. Add in testing of driver
namelist. Work on SSP test and add SPL testcase. Add option to
Namelist.pm->merge_nl to die on conflict. change notes for Namelist.pm
to be "!#" instead of "#!". Fix namelist test to work. Remove messing
with drv_flds_in file in preview_namelists.

2) Changes to deposition tables in seq_drydep_mod for additional VOC type
species.

3) Remove inputdata files from usermods that referenced glade so
that tests will pass on tests outside NCAR.

4) Rework the vertical gradient calculator for lnd to glc remapping in a
few ways:

(a) Adds a limiter to the vertical gradient calculator
(vertical_gradient_calculator_2nd_order), restricting gradients such
that the values at the interface between two elevation classes must
lie in between the mean values in those two elevation classes. Bill
Lipscomb provided the initial implementation for this change.

(b) Significantly reworks the API and internals of the vertical gradient
calculator, as well as related unit tests.

(c) Adds a factory module to create vertical gradient calculators -
making the vertical gradient calculator's constructor independent of
attribute vectors

(c) Adds a script for plotting vertical gradients

Changes (b), (c) and (d) were not necessary to accomplish (a1). Rather,
these changes and additions were made in order to support the
implementation of a new, more complex vertical gradient calculator
(which ensured continuity at elevation class interfaces via a matrix
solve for the gradients). The results of this new vertical gradient
calculator were not as good as we had hoped, so Bill Sacks backed that out -
but kept in place the changes in (b), (c) and (d) because these seemed
to be an improvement to the design.

Changes answers for all compsets with an active GLC (i.e., with CISM). 
This only feeds back on the climate if two-way evolution is on
otherwise, this is a diagnostic-only change. 

5) Also fixed problems with the short term archiver, etc that were needed
to address clm45_aux testing issues

Test suite: scripts_regression_tests
Test baseline: NA
Test namelist changes: NA
Test status: all passed

Fixes: [CIME Github issue #]

User interface changes?: None

Code review:

### CODE REVIEW

File: cime_config/cesm/allactive/testlist_allactive.xml [hunk 1]
```diff
@@ -400,4 +384,256 @@
       </options>
     </machines>
   </test>
+=======
```
Comment (jgfouca) [inline cime_config/cesm/allactive/testlist_allactive.xml]: These ==== shoudn't be here.

File: driver_cpl/bld/build-namelist [hunk 2]
```diff
@@ -176,7 +176,7 @@ EOF
 EOF
 
 # The Build::NamelistDefinition module provides utilities to validate that the output
-# namelists are consistent with the namelist definition file
+# namelists are consistent with the namelist definition files
```
Comment (jgfouca) [inline driver_cpl/bld/build-namelist]: Isn't the perl version of build-namelist not used anymore?

File: driver_cpl/cime_config/testdefs/testlist_drv.xml [hunk 3]
```diff
@@ -672,4 +654,195 @@
       </machine>
     </machines>
   </test>
+=======
```
Comment (jgfouca) [inline driver_cpl/cime_config/testdefs/testlist_drv.xml]: ====== again

File: scripts/Tools/compare_namelists.pl [hunk 4]
```diff
@@ -23,14 +23,18 @@
 my $cnt2=$#basenml;
 my $shiftbl=1;
 foreach $line (@nlfile){
```
Comment (jgfouca) [inline scripts/Tools/compare_namelists.pl]: I don't think this is used anymore.

File: utils/python/CIME/case_st_archive.py [hunk 5]
```diff
@@ -178,7 +178,8 @@ def get_histfiles_for_restarts(case, archive, archive_entry, restfile):
         rundir = case.get_value("RUNDIR")
         cmd = "ncdump -v %s %s " %(rest_hist_varname, os.path.join(rundir, restfile))
         rc, out, error = run_cmd(cmd, ok_to_fail=True)
-        expect(rc == 0, "Command %s failed rc=%d\nout=%s\nerr=%s" %(cmd, rc, out, error))
+        if rc != 0:
```
Comment (jgfouca) [inline utils/python/CIME/case_st_archive.py]: It's only a warning if ncdump fails?

**Comment by rljacob** [comment]
What is this new driver_cpl/bld/tesdir directory for?

**Comment by rljacob** [comment]
Is this BFB in the coupler?

**Comment by jgfouca** [comment]
@mvertens I'm not qualified to review most of this very large PR, but I did my best. A couple changes needed then I'll merge.

**Comment by ekluzek** [comment]
Hi Rob

Currently the driver build-namelist can not be exercised outside of the
context of setting up a case with the cime scripts. You can't just run it
standalone and see if it gives you a reasonable namelist. So I created a
test directory so that you could at least run the build-namelist on it's
own and look at the resulting namelist. This makes working with the driver
build-namelist faster and easier. It doesn't verify the namelist is
correct, nor does it check that the logic in build-namelist is working
correctly, but it does allow for a simple straightforward check.

*Erik Kluzek, (CGD at NCAR)*

*National Center for Atmospheric Research*

*Boulder CO, *

*(off) (303)497-1326 (fax) (303)497-1348*

*(skype) ekluzek    (cell) (303)859-0183*

*Pronouns: he/his/him*

*------------------ Home page ------------------------*

*          http://www.cgd.ucar.edu/~erik <http://www.cgd.ucar.edu/~erik>*

*          https://staff.ucar.edu/users/erik
<https://staff.ucar.edu/users/erik>*

*!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!*

File: driver_cpl/bld/build-namelist (hunk 2 above)
Comment (mvertens) [inline driver_cpl/bld/build-namelist]: Yes - the perl version of build-namelist is still being used (I wish it were not - but its going to be a bit of a larger refactoring effort to move this to python).

File: driver_cpl/cime_config/testdefs/testlist_drv.xml (hunk 3 above)
Comment (mvertens) [inline driver_cpl/cime_config/testdefs/testlist_drv.xml]: Thanks for catching this. The merge added the old schema to the new. Its now fixed.

File: scripts/Tools/compare_namelists.pl (hunk 4 above)
Comment (mvertens) [inline scripts/Tools/compare_namelists.pl]: Do you mean that compare_namelists.pl is not used anymore?

File: utils/python/CIME/case_st_archive.py (hunk 5 above)
Comment (mvertens) [inline utils/python/CIME/case_st_archive.py]: Yes. This variable is needed to tell you the variable name on the restart file that contains settings for what history files need to be in your exeroot on restart. What I did not catch earlier is that some components (like cam, clm) contain history restart files - and that variable will not be on those files. So at this point, I don't want to change the scheme of the the config_archive.xml  - so as a result we are just printing out a warning and continuing. I caught this in running the aux_clm45 tests with the latest cime.

File: scripts/Tools/compare_namelists.pl (hunk 4 above)
Comment (jgfouca) [inline scripts/Tools/compare_namelists.pl]: Yes, compare_namelists.pl should not be used anymore and should be deleted.

**Comment by mvertens** [comment]
Update the PR.

File: cime_config/cesm/allactive/testlist_allactive.xml (hunk 1 above)
Comment (mvertens) [inline cime_config/cesm/allactive/testlist_allactive.xml]: Agreed. This is fixed.

**Comment by jgfouca** [comment]
@mvertens , I still the ====== in a couple files. Also, the perl compare_namelists should be deleted.

**Comment by mvertens** [comment]
I only found ==== in testlist_allactive.xml. I removed the one in testlist_drv.xml and also removed compare_namelists.pl. I am not sure why you are still seeing those. I just cloned a clean copy of branch and they are not there.

**Comment by jgfouca** [comment]
@mvertens the only update that I see to this branch since it was opened was a small commit by @quantheory (Fix machines error in detecting threaded builds.) Are you sure you're pushing your fixes to the right branch?

**Comment by mvertens** [comment]
@jgfouca I think I fixed things.This was my fault - I thought I was pushing successfully - but it actually was giving me an error message which I did not bother to look at. Both Sean's and my changes to address your comments should be in the latest PR update.

**Comment by mvertens** [comment]
@jgfouca Please hold off on merging this PR - I am find a problem in some of the aux cesm tests with this branch

**Comment by mvertens** [comment]
@jgfouca I have implemented fixes and have successfully run the regression test. So please go ahead and resolve this PR if you are comfortable with everything.

**Comment by mvertens** [comment]
I did just push my latest changes.

**Comment by jgfouca** [comment]
Looks good, thanks.

**Pull request merged**